workitem/workitem_repository.go

```diff
@@ -93,6 +112,110 @@ func (r *GormWorkItemRepository) Delete(ctx context.Context, ID string) error {
 	return nil
 }
 
+// Reorder places the to-be-reordered workitem(s) above the input workitem.
+// The order of workitems are spaced by a factor of 1000.
+// The new order of workitem := (order of previousitem + order of nextitem)/2
+// Version must be the same as the one int the stored version
+func (r *GormWorkItemRepository) Reorder(ctx context.Context, beforeID string, wi app.WorkItem) (*app.WorkItem, error) {
+	var order float64
+	res := WorkItem{}
+	beforeItem := WorkItem{}
+	afterItem := WorkItem{}
+	id, err := strconv.ParseUint(wi.ID, 10, 64)
+	if err != nil || id == 0 {
+		return nil, errors.NewNotFoundError("work item", wi.ID)
+	}
+
+	tx := r.db.First(&res, id)
+	if tx.RecordNotFound() {
+		return nil, errors.NewNotFoundError("work item", wi.ID)
+	}
+	if tx.Error != nil {
+		return nil, errors.NewInternalError(err.Error())
+	}
+	if res.Version != wi.Version {
+		return nil, errors.NewVersionConflictError("version conflict")
+	}
+
+	wiType, err := r.wir.LoadTypeFromDB(ctx, wi.Type)
+	if err != nil {
+		return nil, errors.NewBadParameterError("Type", wi.Type)
+	}
+
+	if beforeID != "" {
+		beforeId, err := strconv.ParseUint(beforeID, 10, 64)
+		if err != nil || beforeId == 0 {
+			return nil, errors.NewNotFoundError("work item", string(beforeId))
+		}
+		tx = r.db.First(&beforeItem, beforeId)
+		if tx.RecordNotFound() {
+			return nil, errors.NewNotFoundError("work item", string(beforeId))
+		}
+		if tx.Error != nil {
+			return nil, errors.NewInternalError(err.Error())
+		}
+		beforeOrder, err := strconv.ParseFloat(fmt.Sprintf("%v", beforeItem.Position), 64)
+		if err != nil {
+			return nil, errors.NewBadParameterError("data.attributes.order", res.Position)
+		}
+		tx2 := r.db.Where("position < ?", beforeItem.Position).Order("position desc", true).Last(&afterItem)
+		if afterItem.ID == 0 {
+			// The item is moved to first position
+			order = (0 + beforeOrder) / 2
+		} else {
+			afterOrder, err := strconv.ParseFloat(fmt.Sprintf("%v", afterItem.Position), 64)
+			if err != nil {
+				return nil, errors.NewBadParameterError("data.attributes.order", res.Position)
+			}
+			if tx2.RecordNotFound() {
+				return nil, errors.NewNotFoundError("work item", string(beforeId))
+			}
+			if tx2.Error != nil {
+				return nil, errors.NewInternalError(err.Error())
+			}
+			order = (beforeOrder + afterOrder) / 2
+		}
+	} else {
+		// the item is moved at last position
+
+		tx2 := r.db.Order("position desc", true).Last(&afterItem)
+		if tx2.RecordNotFound() {
+			return nil, errors.NewNotFoundError("work item", string(afterItem.ID))
+		}
+		if tx2.Error != nil {
+			return nil, errors.NewInternalError(err.Error())
+		}
+		afterOrder, _ := strconv.ParseFloat(fmt.Sprintf("%v", afterItem.Position), 64)
+		order = afterOrder + 1000
```

Comment:
@maxandersen I dont think that there could be any case where two workitems will have same order because:
[1] Order of new workitem is (highest-order-in-db + 1000) which is unique
[2] While reordering, 
Order of the reorder workitem is obtained by dividing the order of the two 'closest' workitems, so the order will always be a value between the range of those two 'closest' workitems. It will never be **equal** to the order of those two closest workitems. 

(By two 'closest' workitems, I mean the workitems below and above which the reorder workitem is going to be placed.)